### PR TITLE
cpu/samd5x: RTC: add timeout to spurious tamper event

### PR DESCRIPTION
**Pull request by benpicco**
### Contribution description

Tamper events are used to wake the CPU from Deep Sleep. On enabling, a spurious tamper event is generated, so we block on a mutex that gets unlocked by the tamper event callback before we enter sleep in this case.

Errata 2.17.4 says:

> Upon enabling the RTC tamper detection feature, a false tamper detection *can* be reported by the RTC.

It turns out that this spurious event is not always generated.
If RTC alarm is used and the CPU was previously woken from hibernate by RTC, it *can* happen that the false tamper event is *not* generated.

In this case, we will block indefinitely on the mutex.

To solve this, add a timeout to the event.
Also poll the event instead of using a mutex, as we have already set `PM->SLEEPCFG.bit.SLEEPMODE` at this point.

### Testing procedure

This has proven difficult / non deterministic.

I managed to reproduce the issue by introducing a `init_int_rtc` function to `tests/periph_gpio` that will configure the tamper pin, set an RTC alarm and then put the CPU to hibernate. A GPIO of a second board was connected to the tamper pin with a GPIO set to output using `tests/periph_gpio`.

After a few iterations where the CPU was woken by RTC, the tamper false event would not be generated anymore before entering sleep and execution would get stuck.

```patch
--- a/tests/periph_gpio/main.c
+++ b/tests/periph_gpio/main.c
@@ -25,6 +25,7 @@
 #include "shell.h"
 #include "benchmark.h"
 #include "periph/gpio.h"
+#include "periph/rtc.h"
 
 #define BENCH_RUNS_DEFAULT      (1000UL * 100)
 #define IRQ_TIMEOUT_US          (1000UL)
@@ -187,6 +188,25 @@ static int enable_int(int argc, char **argv)
 
     return 0;
 }
+
+static int init_int_rtc(int argc, char **argv)
+{
+    unsigned secs = 5;
+
+    if (init_int(argc, argv)) {
+        return 1;
+    }
+
+    struct tm now;
+    rtc_get_time(&now);
+    now.tm_sec += secs;
+    rtc_set_alarm(&now, cb, NULL);
+
+    printf("will hibernate for %u sec\n", secs);
+    pm_set(0);
+
+    return 0;
+}
 #endif
 
 static int cmd_read(int argc, char **argv)
@@ -409,6 +429,7 @@ static const shell_command_t shell_commands[] = {
     { "init_od_pu", "init as output (open-drain with pull-up)", init_od_pu },
 #ifdef MODULE_PERIPH_GPIO_IRQ
     { "init_int", "init as external INT w/o pull resistor", init_int },
+    { "init_int_rtc", "init as external INT w/o pull resistor", init_int_rtc },
     { "enable_int", "enable or disable gpio interrupt", enable_int },
 #endif
     { "read", "read pin status", cmd_read },
```

I also added a `timeout` debug output when `timeout == 0` after the loop and a `TAMPER` print output to `_isr_tamper()`. (did not commit these)

This produced the following results (with this patch applied):

```
2020-12-04 00:21:11,608 # main(): This is RIOT! (Version: 2021.01-devel-1349-g9d46bc)
2020-12-04 00:21:11,611 # GPIO peripheral driver test
2020-12-04 00:21:11,611 # 
2020-12-04 00:21:11,617 # In this test, pins are specified by integer port and pin numbers.
2020-12-04 00:21:11,622 # So if your platform has a pin PA01, it will be port=0 and pin=1,
2020-12-04 00:21:11,626 # PC14 would be port=2 and pin=14 etc.
2020-12-04 00:21:11,626 # 
2020-12-04 00:21:11,631 # NOTE: make sure the values you use exist on your platform! The
2020-12-04 00:21:11,636 #       behavior for not existing ports/pins is not defined!
2020-12-04 00:21:14,045 #  init_int_rtc 1 0 1
2020-12-04 00:21:14,055 # GPIO_PIN(1, 0) successfully initialized as ext int
2020-12-04 00:21:14,060 # will hibernate for 10 sec
2020-12-04 00:21:14,061 # enable tamper
2020-12-04 00:21:14,062 # TAMPER # false tamper event
2020-12-04 00:21:14,063 # tamper enabled
2020-12-04 00:21:24,470 # main(): This is RIOT! (Version: 2021.01-devel-1349-g9d46bc) # woken by RTC
2020-12-04 00:21:24,472 # GPIO peripheral driver test
2020-12-04 00:21:24,472 # 
2020-12-04 00:21:24,478 # In this test, pins are specified by integer port and pin numbers.
2020-12-04 00:21:24,483 # So if your platform has a pin PA01, it will be port=0 and pin=1,
2020-12-04 00:21:24,487 # PC14 would be port=2 and pin=14 etc.
2020-12-04 00:21:24,487 # 
2020-12-04 00:21:24,492 # NOTE: make sure the values you use exist on your platform! The
2020-12-04 00:21:24,497 #       behavior for not existing ports/pins is not defined!
2020-12-04 00:21:29,053 #  init_int_rtc 1 0 1
2020-12-04 00:21:29,063 # GPIO_PIN(1, 0) successfully initialized as ext int
2020-12-04 00:21:29,067 # will hibernate for 10 sec
2020-12-04 00:21:29,068 # enable tamper
2020-12-04 00:21:29,069 # timeout # no false tamper event
2020-12-04 00:21:29,071 # tamper enabled
2020-12-04 00:21:33,113 # TAMPER # woken by other board
2020-12-04 00:21:33,118 # main(): This is RIOT! (Version: 2021.01-devel-1349-g9d46bc)
2020-12-04 00:21:33,120 # GPIO peripheral driver test
2020-12-04 00:21:33,120 # 
2020-12-04 00:21:33,126 # In this test, pins are specified by integer port and pin numbers.
2020-12-04 00:21:33,132 # So if your platform has a pin PA01, it will be port=0 and pin=1,
2020-12-04 00:21:33,135 # PC14 would be port=2 and pin=14 etc.
2020-12-04 00:21:33,135 # 
2020-12-04 00:21:33,140 # NOTE: make sure the values you use exist on your platform! The
2020-12-04 00:21:33,146 #       behavior for not existing ports/pins is not defined!
```
(it took several tries to get the `timeout` event, usually I get the false `TAMPER` event)

### CODE REVIEW

File: cpu/sam0_common/periph/rtc_rtt.c [hunk 1]
```diff
@@ -337,31 +337,36 @@ int rtc_tamper_register(gpio_t pin, gpio_flank_t flank)
...
-        tamper_cb.cb  = _unlock;
-        tamper_cb.arg = &m;
-    }
 
-    /* enable tamper detect as wake-up source */
-    RTC->MODE0.INTENSET.bit.TAMPER = 1;
+        /* If an RTC alarm happened before, the spurious tamper
+         * event is sometimes generated.
+         * Tamper event must happen within one RTC clock period. */
+        unsigned timeout = CLOCK_CORECLOCK / 32768;
```
Comment (benpicco) [inline cpu/sam0_common/periph/rtc_rtt.c:354]: We could further divide the timeout value here as the `!RTC->MODE0.INTFLAG.bit.TAMPER && --timeout` check takes more than once cycle.

File: cpu/sam0_common/periph/rtc_rtt.c [hunk 2]
```diff
@@ -337,31 +336,42 @@ int rtc_tamper_register(gpio_t pin, gpio_flank_t flank)
...
     if (RTC->MODE0.TAMPCTRL.reg & RTC_TAMPCTRL_TAMLVL_Msk) {
-        mutex_lock(&m);
-        tamper_cb.cb  = _unlock;
-        tamper_cb.arg = &m;
-    }
 
-    /* enable tamper detect as wake-up source */
-    RTC->MODE0.INTENSET.bit.TAMPER = 1;
+        /* If an RTC alarm happened before, the spurious tamper
+         * event is sometimes generated.
```
Comment (benpicco) [inline cpu/sam0_common/periph/rtc_rtt.c:351]:
```suggestion
         * event is sometimes not generated.
```

File: cpu/sam0_common/periph/rtc_rtt.c (hunk 1 above)
Comment (maribu) [reply cpu/sam0_common/periph/rtc_rtt.c:353]: Better safe than sorry.

**Comment by maribu** [comment]
I don't have the hardware to test, but I trust your testing. In addition, the observed behavior matches what the errata sheet says and the code is very straightforward.

**Comment by maribu** [comment]
Care to squash in the missing `not`?

**Comment by benpicco** [comment]
Also rebased onto master to accommodate the updates to the CI

**Review by maribu — APPROVED** [review]
ACK

**Comment by benpicco** [comment]
Thank you for the review!
I tested this on `same54-xpro`

**Pull request merged**